File: backend_go/internal/service/game_service.go
package service

import (
	"gestory-backend/internal/dto"
	"gestory-backend/internal/models"
	"gestory-backend/internal/repository"
	"time"
)

type GameService interface {
	SubmitScore(req dto.SubmitGameScoreRequest) (*dto.SubmitGameScoreResponse, error)
	GetLeaderboard(limit int) ([]dto.LeaderboardItem, error)
}

type gameService struct {
	gameRepo repository.GameRepository
}

func NewGameService(gameRepo repository.GameRepository) GameService {
	return &gameService{gameRepo: gameRepo}
}

func (s *gameService) SubmitScore(req dto.SubmitGameScoreRequest) (*dto.SubmitGameScoreResponse, error) {
	scoreModel := models.GameScore{
		PlayerName:        req.PlayerName,
		Score:             req.FinalScore,
		QuestionsAnswered: req.QuestionsAnswered,
		CorrectAnswers:    req.CorrectAnswers,
		Accuracy:          req.Accuracy,
		TimeSpent:         req.TimeSpent,
		CompletedAt:       time.Now(),
	}

	err := s.gameRepo.Create(&scoreModel)
	if err != nil {
		return nil, err
	}

	return &dto.SubmitGameScoreResponse{
		ID:         scoreModel.ID.String(),
		PlayerName: scoreModel.PlayerName,
		FinalScore: scoreModel.Score,
		CreatedAt:  scoreModel.CompletedAt,
	}, nil
}

func (s *gameService) GetLeaderboard(limit int) ([]dto.LeaderboardItem, error) {
	scores, err := s.gameRepo.GetTopScores(limit)
	if err != nil {
		return nil, err
	}

	items := make([]dto.LeaderboardItem, 0, len(scores))
	for i, score := range scores {
		items = append(items, dto.LeaderboardItem{
			Rank:       i + 1,
			Score:      score.Score,
			PlayerName: score.PlayerName,
			Date:       score.CompletedAt,
		})
	}

	return items, nil
}

File: gestory/src/services/gameService.ts
/**
 * Game Service
 * Real backend via apiClient.
 */

import { apiClient } from "./api";
import type { ApiResponse, GameScore, QuizQuestionsResponse } from "@/types";

// Backend has: GET /api/v1/game/questions -> returns QuizQuestionsResponse shape
export async function getGameQuestions(): Promise<ApiResponse<QuizQuestionsResponse>> {
  return apiClient.get(`/api/v1/game/questions`);
}

export async function submitGameScore(scoreData: {
  playerName: string;
  finalScore: number;
  questionsAnswered: number;
  correctAnswers: number;
  accuracy: number;
  timeSpent: number;
}): Promise<ApiResponse<GameScore>> {
  return apiClient.post(`/api/v1/game/submit`, scoreData);
}

export async function getGameLeaderboard(
  limit: number = 10
): Promise<ApiResponse<Array<{ rank: number; score: number; playerName: string; date: string }>>> {
  return apiClient.get(`/api/v1/game/leaderboard?limit=${limit}`);
}

export const gameService = {
  getGameQuestions,
  submitGameScore,
  getGameLeaderboard,
};

